File: 0282-expression-add-operators/0282-expression-add-operators.java
class Solution {
    public List<String> addOperators(String num, int target) {
        List<String> ans=new ArrayList<>();
        solve(0,num,target,ans,"",0L,0L);
        return ans;
    }

    void solve(int ind,String num,int target,List<String> ans,String path,long total,long prev){
    if (ind==num.length())
{        if(total==target){
            ans.add(path);
        }
        return;
        }

        for(int i=ind;i<num.length();i++){
             if (i > ind && num.charAt(ind) == '0') break;

            String part = num.substring(ind, i + 1);
            long curr = Long.parseLong(part);
            if(ind==0){
                solve(i+1,num,target,ans,part,curr,curr);
            }else{
                 solve(i + 1, num, target, ans,
                      path + "+" + part,
                      total + curr,
                      curr);
                solve(i + 1, num, target, ans,
                      path + "-" + part,
                      total - curr,
                      -curr);   
                solve(i + 1, num, target, ans,
                      path + "*" + part,
                      total - prev + (prev * curr),
                      prev * curr);
            }

        }
    }
}

File: 0282-expression-add-operators/0282-expression-add-operators.py
class Solution:
    def __init__(self):
        self.a=[]
    def addOperators(self, num: str, target: int) -> List[str]:
        self.solve(num,target,0,0,0,"")
        return self.a
    def solve(self,num,target,ind,total,prev,s):
        if ind==len(num):
            if target==total:
                self.a.append(s)
            return
        for i in range(ind,len(num)):
            if i!=ind and num[ind]=='0':
                break

            st=num[ind:i+1]
            n=int(st)
            if(ind==0):
                self.solve(num,target,i+1,n,n,s+st)
            else:
                self.solve(num,target,i+1,total+n,n,s+'+'+st)
                self.solve(num,target,i+1,total-n,n,s+'-'+st)
                self.solve(num,target,i+1,total-prev+(n*prev),n,s+'*'+st)
            


        